Devicetree: edtlib: clarify the possibly confusing content of _VENDOR_PREFIX_ALLOWED

dottspina opened the pull request:
This PR groups a few commits related to the vendor prefixes support in the edtlib API, one of which may be a bug fix.

### Fix typo in vendor prefixes

This one should be straightforward: there's a typo where the compatible string `gpio-keys` is misspelled `gpio-key` in the *grandfathered in by Linux* compatible strings set.

Reference: [gpio-keys.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/input/gpio-keys.txt) 

### Complement allowed vendor prefixes

Add a few vendor prefixes *grandfathered in by Linux*:

- `fixed-partitions` ([fixed-partitions.yaml](https://www.kernel.org/doc/Documentation/devicetree/bindings/mtd/partitions/fixed-partitions.yaml))
- `leds-gpio.txt` ([leds-gpio.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/leds-gpio.txt))
- `mmio-sram` ([sram.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/sram/sram.txt))
- `pwm-leds` ([leds-pwm.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/leds-pwm.txt))
 
### Fix *vendor prefixes validation*

When initialized with a vendor prefixes dictionary, an EDT model will permit to retrieve vendor descriptions with the public `EDT.compat2vendor` property.

Additionally, when this vendors support is enabled, the `EDT._init_luts()` function seems to validate compatible strings: for all nodes but the devicetree root, compatible strings should either include a vendor defined in the `vendor-prefixes.txt` dictionary in use, or be *grandfathered in by Linux*.

The issue is that:
- a compatible string is "grandfathered in by Linux" if present in the `_VENDOR_PREFIX_ALLOWED` set, which indeed contains compatible values, **NOT** vendor prefixes
- the validation happens **ONLY** when the compatible string starts with a vendor prefix followed by a comma, and there `_VENDOR_PREFIX_ALLOWED` would be searched for a vendor prefix it definitely cannot contain

IIUC the intent, `EDT._init_luts()` should:
- validate compatible strings whenever the vendors support is enabled (`EDT._vendor_prefixes` is not `None`)
- search the `EDT._vendor_prefixes` dictionary (aka `vendor-prefixes.txt`) for vendors when the compatible string actually starts with a vendor prefix
- search the *grandfathered in by Linux* set to validate compatible strings that do not start with a vendor prefix

Hope this helps.
Thanks.


--
chris

dottspina commented:
This PR breaks the `test_deepcopy()` unit test in `test_edtlib.py`:

```
msg = "node '/in-dir-1' compatible 'in-dir-1' has unknown vendor prefix 'None'"

    def _err(msg) -> NoReturn:
>       raise EDTError(msg)
E       devicetree.edtlib.EDTError: node '/in-dir-1' compatible 'in-dir-1' has unknown vendor prefix 'None'
```

The related DTS file (`test-multidir.dts`) is quite simple:

``` 
/ {
	in-dir-1 {
		compatible = "in-dir-1";
	};
	in-dir-2 {
		compatible = "in-dir-2";
	};
};
```

If think this illustrates quite well the (probably juste mine) confusion regarding when EDT should complain about a compatible string with unknown vendor. 

To me the unit test rightly fails:

- the "in-dir-1" compatible string is indeed not defined in the provided `vendor_prefixes` dictionnary
- the "in-dir-1" compatible string is not *grandfathered in by Linux* (i.e. not in `_VENDOR_PREFIX_ALLOWED`)
- IIUC it's precisely a situation when we expect EDT to complain (here not about an undefined vendor prefix, but about a compatible string that neither matches a vendor in `vendor_prefixes`, nor a compatible value in the *grandfathered in by Linux* set)    

Thus, IIUC (still) we should consider this PR and patch the unit test, something like bellow (untested):  

``` diff
@@ -730,7 +730,11 @@ def test_deepcopy():
                          default_prop_types=False,
                          support_fixed_partitions_on_any_bus=False,
                          infer_binding_for_paths=['/test-node'],
-                         vendor_prefixes={'test-vnd': 'A test vendor'},
+                         vendor_prefixes={
+                            'test-vnd': 'A test vendor',
+                            'in-dir-1': 'Work-around for test_deepcopy()',
+                            'in-dir-2': 'Work-around for test_deepcopy()',
+                            },
                          werror=True)
         edt_copy = deepcopy(edt)

@@ -775,7 +779,11 @@ def test_deepcopy():
     assert not edt_copy._fixed_partitions_no_bus
     assert edt_copy._infer_binding_for_paths == set(["/test-node"])
     assert edt_copy._infer_binding_for_paths is not edt._infer_binding_for_paths
-    assert edt_copy._vendor_prefixes == {"test-vnd": "A test vendor"}
+    assert edt_copy._vendor_prefixes == {
+        "test-vnd": "A test vendor",
+        'in-dir-1': 'Work-around for test_deepcopy()',
+        'in-dir-2': 'Work-around for test_deepcopy()',
+    }
     assert edt_copy._vendor_prefixes is not edt._vendor_prefixes
     assert edt_copy._werror
     test_equal_but_not_same("_compat2binding", equal_key2path)
```

I've not pushed the above patch, we first have to confirm or refute my (mis)understanding of this *thing*.

mbolivar-ampere reviewed (changes requested):
These are meant to be vendor prefixes, not compatible strings. Where is the impression that they are compatible strings coming from? Please see https://lxr.linux.no/#linux+v6.0.9/Documentation/devicetree/bindings/vendor-prefixes.yaml#L16

dottspina commented:
> These are meant to be vendor prefixes, not compatible strings. Where is the impression that they are compatible strings coming from? 

Well, subjectively, strings like "gpio-fan" or "gpio-keys" really look like models rather than manufacturers, and failing to grep anything like `gpio-fan,` (with a comma) in the Zephyr binding and DTS files achieved to wrongly convince me they never appear as *vendors*. 

Sorry, I entirely missed the Linux `patternProperties` thing you're kindly pointing me toward. 

> Please see https://lxr.linux.no/#linux+v6.0.9/Documentation/devicetree/bindings/vendor-prefixes.yaml#L16

Thanks. This link does not work for me but finding the vendor-prefixes.yaml file was straightforward (e.g. [here](https://www.kernel.org/doc/Documentation/devicetree/bindings/vendor-prefixes.yaml)).

According to the `patternProperties` definition, I've reverted all previous commits in this PR. Sorry for this. 

While I was at it, I've looked through these prefixes usage by the Linux Devicetree: AFAICT, they may appear as the manufacturer in property names and in values of the "model" standard property, but never as the manufacturer in compatible strings.

Using them in `EDT._init_luts()` to allow additional manufacturers when validating compatible strings hence sounds worthless, and remains a little confusing to me: `vendor in _VENDOR_PREFIX_ALLOWED` is unlikely to happen.

This isn't incorrect in any way, though: do not hesitate to close this PR (a single commit remains, that adds a few comments *reminding* where these prefixes come from, and where they appear as manufacturers in the Linux kernel). 

Thanks.

mbolivar-ampere commented:
How strange that the link is not working for you. I was able to load it just now myself. At any rate, I am glad you were able to find the file.

Just to be clear, it sounds like you are saying we shouldn't be allowing these as special cases in compatible properties, because Linux isn't doing that, either, and so therefore the suggested patch should be something along the lines of this?

```patch
diff --git a/scripts/dts/python-devicetree/src/devicetree/edtlib.py b/scripts/dts/python-devicetree/src/devicetree/edtlib.py
index faf51e58be..bc8bbd51c5 100644
--- a/scripts/dts/python-devicetree/src/devicetree/edtlib.py
+++ b/scripts/dts/python-devicetree/src/devicetree/edtlib.py
@@ -2247,8 +2247,7 @@ class EDT:
 
                     # As an exception, the root node can have whatever
                     # compatibles it wants. Other nodes get checked.
-                    elif node.path != '/' and \
-                       vendor not in _VENDOR_PREFIX_ALLOWED:
+                    elif node.path != '/':
                         if self._werror:
                             handler_fn: Any = _err
                         else:
@@ -3214,14 +3213,3 @@ _DEFAULT_PROP_SPECS: Dict[str, PropertySpec] = {
     name: PropertySpec(name, _DEFAULT_PROP_BINDING)
     for name in _DEFAULT_PROP_TYPES
 }
-
-# A set of vendor prefixes which are grandfathered in by Linux,
-# and therefore by us as well.
-_VENDOR_PREFIX_ALLOWED: Set[str] = set([
-    "at25", "bm", "devbus", "dmacap", "dsa",
-    "exynos", "fsia", "fsib", "gpio-fan", "gpio-key", "gpio", "gpmc",
-    "hdmi", "i2c-gpio", "keypad", "m25p", "max8952", "max8997",
-    "max8998", "mpmc", "pinctrl-single", "#pinctrl-single", "PowerPC",
-    "pl022", "pxa-mmc", "rcar_sound", "rotary-encoder", "s5m8767",
-    "sdhci", "simple-audio-card", "st-plgpio", "st-spics", "ts",
-])
```

dottspina commented:
> How strange that the link is not working for you. I was able to load it just now myself. At any rate, I am glad you were able to find the file.

I personally still get  `SSL_ERROR_RX_RECORD_TOO_LONG`, but I'm quite sure this relates to some recent issues with SSL in Firefox that are specific to RHEL9 (e.g. [2228948](https://bugzilla.redhat.com/show_bug.cgi?id=2228948)).
 
> Just to be clear, it sounds like you are saying we shouldn't be allowing these as special cases in compatible properties, because Linux isn't doing that, either, 

Not quite.

IIUC, the binding files in Linux exist mainly for documentation purpose, and, since documentation is important, `checkpatch.pl` will verify "Any string used in a compatible property must be documented as to what it indicates" ([usage-model.rst](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/usage-model.rst)): there, a manufacturer is deemed documented (and thus valid) when found in `vendor-prefixes.yaml`.

So it's not that Linux isn't allowing these prefixes in compatible values: `checkpatch.pl` implicitly does (they *are* in the `vendor-prefixes.yaml` file).

It just happens that the "Prefixes which are not vendors" do not define (document) manufacturers that will appear in compatible properties. Besides, `checkpatch.pl` still succeeds after they're removed from `vendor-prefixes.yaml` (but rightly fails when an actual vendor's definition is removed).

> and so therefore the suggested patch should be something along the lines of this?

Basically, yes, that would:

- remove an unreachable code path
- prevent readers from believing these strings are actually valid manufacturers that may appear in compatible properties    

But, assuming the rationale for adding these "Prefixes which are not vendors" to edtlib was some kind of loyalty to Linux, we may want to nonetheless document them in some way.
 
Adding ad-hoc entries to `vendor-prefixes.txt` seems appealing: that would incidentally *promote* these prefixes to valid vendors, but isn't that what Linux does, equally incidentally ?

This way `EDT._init_luts()` will eventually behave exactly the same as before despite the removed code: the special cases are just no longer special (Linux does not treat them as special, either).

I guess `vendor-prefixes.txt` is a sensitive file and will wait for your feedback before I draft the changes.

Thanks.

dottspina committed:
devicetree: edtlib: prefixes which are not vendors are NOT vendors

In Linux, checkpatch.pl relies on the vendor-prefixes.yaml file
to validate manufacturers in compatible strings.
In addition to the vendors defined in vendor-prefixes.txt,
the YAML file includes expressions for "prefixes which are not vendors":
these expressions do NOT define special manufacturers that may appear
in compatible strings, and are never involved as such in DTS files.
We can rather see them as bulk-definitions of JSON/YAML properties
suitable for the dt-schema tools.

OTHO, in Zephyr, checkpatch.pl relies on the vendor-prefixes.txt file,
which does not include these additional prefixes, but edtlib.EDT adds
them as hard-coded special values.

This is confusing, if not incorrect:

- the fact that edtlib.EDT (and thus its client code in the
zephyr/scripts directory) actually allows these vendors
in compatible strings is buried in the source code
- checkpatch.pl (with vendor-prefixes.txt) in Zephyr behaves neither like
checkpatch.pl (with vendor-prefixes.yaml) in Linux, nor like edtlib.EDT
(with _VENDOR_PREFIX_ALLOWED)
- Zephyr should not treat these "prefixes which are not vendors" as
valid manufacturers in compatible strings to begin with

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

dottspina commented:
I put this in a new comment to make it clear.

> But, assuming the rationale for adding these "Prefixes which are not vendors" to edtlib was some kind of loyalty to Linux, we may want to nonetheless document them in some way.
> 
> Adding ad-hoc entries to `vendor-prefixes.txt` seems appealing: that would incidentally _promote_ these prefixes to valid vendors, but isn't that what Linux does, equally incidentally ?

No, I was wrong, we do NOT want to *incidentally* allow these "prefixes which are not vendors" as valid manufacturers in compatible strings. The sane approach is to entirely remove them from edtlib, without having to document them *somewhere else* in Zephyr.

Practically, this means we should just apply the patch @mbolivar-ampere has already drafted. I've updated this PR accordingly.

In the commit message I wrote that, besides in checkpatch.pl, the vendor-prefixes-yaml file in Linux may also be used to bulk-define properties when validating DTS files with the dt-schema tools: could someone please clarify whether this is correct or not ?

Thanks.

mbolivar-ampere reviewed (approved):
Thanks for the analysis. This seems reasonable to me.
